### ISSUE TITLE
21x — 64x slowdown from #11323 "Always vectorize LinAlg ops for VMVX backend" now that we have fast ukernels

### ISSUE THREAD

**Issue by bjacob** [iree-org/iree#12073]
#11323 has been a great work-around for the hard-to-debug issue #10859 (thanks!!). The downside was to make VMVX even slower, but that understandably wasn't a major concern as long as VMVX was very slow anyway.

Now that we have good microkernels on ARM64, VMVX can in some cases (for now, matmuls) approach peak performance. As VMVX's arithmetic performance gets bumped by 3 or 4 orders of magnitude (that is the gap from interpreting scalar arithmetic ops to calling a good microkernel), the performance trade-offs in work-arounds such as #11323 need to be reevaluated.

To illustrate this, [here](https://gist.github.com/bjacob/06d67b55e78cda31abe109b7a3ffd0b2) are 2 simple matmul testcases, both dynamic shape, one `f32` and one `i8`.

Compile like this,
```
tools/iree-compile --iree-hal-target-backends=vmvx --iree-flow-enable-data-tiling --iree-vmvx-enable-microkernels ~/matmul_f32.mlir -o /tmp/matmul_f32.vmfb
```

Push to device and benchmark like this --- here on the big core of a Moto Edge X30:
```
adb shell taskset 80 /data/local/tmp/iree-benchmark-module --device=local-task --module=/data/local/tmp/matmul_f32.vmfb --function=matmul_f32 --input=1000x1000xf32 --input=1000x1000xf32 --input=1000x1000xf32 --task_topology_group_count=1 --benchmark_min_time=3
```

Results:

Testcase | `main` branch | With #11323 reverted | Speedup from reverting #11323
--- | --- | --- | ---
`matmul_f32` | 1.50 items/s | 31.5 items/s | 21x
`matmul_i8` | 1.55 items/s | 99.1 items/s | 64x

To put this into perspective (since one question that we need to answer here is, does VMVX performance matter?), let's translate this into Gop/s, easy as the matmul size means that each `item` in this benchmark is 2 billion ops, so we just multiply these figures by 2 to get Gop/s:

Testcase | `main` branch | With #11323 reverted
--- | --- | ---
`matmul_f32` | 3.0 Gop/s | 63 Gop/s
`matmul_i8` | 3.1 Gop/s | 198.2 Gop/s

And now let's compare this to peak for this CPU core. It's clocked at 3.0 GHz, does 4 128-bit SIMD arithmetic instructions per cycle. For `f32`, it's vector mul-adds = 8 ops per instruction. For `i8`, it's 2x8x2 matmuls = 64 ops per instruction. So peak is 96 Gop/s for `f32` and 768 Gop/s for `i8`. Dividing the above table by that, we get:

Testcase | `main` branch | With #11323 reverted
--- | --- | ---
`matmul_f32` | 3.1% of peak | 66% of peak
`matmul_i8` | 0.4% of peak | 26% of peak

#12061 is a draft PR essentially reverting #11323. However, #12060 has an e2e testcase that is broken by that --- inadvertently producing a minimized testcase for the originally motivating issue #10859.

**Comment by bjacob** [comment]
Update: once #12099 is merged, we should be able to actually just revert #11323 without inviting back the `alloca` errors that it allowed us to avoid.

**Comment by benvanik** [comment]
Exciting! When it lands we should get some traces of the models hitting the alloca path to see if it's worth optimizing that in the runtime.

**Comment by bjacob** [comment]
Tentative fix is #12061: Instead of just reverting #11323, this keeps and adapts the useful unit test that #11323 introduced, and adds e2e regression tests from the testcases from issues #12060, #10859.
